crypto/store/loader_file.c

```diff
@@ -0,0 +1,1027 @@
+                X509_SIG_free(p8);
+                return NULL;
+            }
+
+            if ((pass = file_get_pass(ui_method, kbuf, PEM_BUFSIZE,
+                                      "PKCS8 decrypt password",
+                                      ui_data)) == NULL) {
+                STOREerr(STORE_F_TRY_DECODE_PKCS8ENCRYPTED,
+                         STORE_R_BAD_PASSWORD_READ);
+                X509_SIG_free(p8);
+                BUF_MEM_free(mem);
+                return NULL;
+            }
+
+            X509_SIG_get0(p8, &dalg, &doct);
+            if (!PKCS12_pbe_crypt(dalg, pass, strlen(pass),
+                                  doct->data, doct->length,
+                                  &new_data, &new_data_len, 0)) {
+                X509_SIG_free(p8);
+                BUF_MEM_free(mem);
+                return NULL;
+            }
+            mem->data = (char *)new_data;
+            mem->max = mem->length = (size_t)new_data_len;
+            X509_SIG_free(p8);
+
+            store_info = store_info_new_DECODED(PEM_STRING_PKCS8INF, mem);
+            if (store_info == NULL) {
+                STOREerr(STORE_F_TRY_DECODE_PKCS8ENCRYPTED,
+                         ERR_R_MALLOC_FAILURE);
+                BUF_MEM_free(mem);
+                return NULL;
+            }
+            return store_info;
+        }
+    }
+
+    return NULL;
+}
+static STORE_FILE_HANDLER PKCS8Encrypted_handler = {
+    "PKCS8Encrypted",
+    try_decode_PKCS8Encrypted
+};
+
+int pem_check_suffix(const char *pem_str, const char *suffix);
+static STORE_INFO *try_decode_PrivateKey(const char *pem_name,
+                                         const char *pem_header,
+                                         const unsigned char *blob,
+                                         size_t len, void **pctx,
+                                         const UI_METHOD *ui_method,
+                                         void *ui_data)
+{
+    STORE_INFO *store_info = NULL;
+    EVP_PKEY *pkey = NULL;
+    const EVP_PKEY_ASN1_METHOD *ameth = NULL;
+
+    if (pem_name != NULL) {
+        if (strcmp(pem_name, PEM_STRING_PKCS8INF) == 0) {
+            PKCS8_PRIV_KEY_INFO *p8inf =
+                d2i_PKCS8_PRIV_KEY_INFO(NULL, &blob, len);
+
+            if (p8inf != NULL)
+                pkey = EVP_PKCS82PKEY(p8inf);
+            PKCS8_PRIV_KEY_INFO_free(p8inf);
+        } else {
+            int slen;
+
+            if ((slen = pem_check_suffix(pem_name, "PRIVATE KEY")) > 0
+                && (ameth = EVP_PKEY_asn1_find_str(NULL, pem_name, slen)) != NULL)
+                pkey = d2i_PrivateKey(ameth->pkey_id, NULL, &blob, len);
+        }
+    } else {
+        int i;
+
+        for (i = 0; i < EVP_PKEY_asn1_get_count(); i++) {
+            ameth = EVP_PKEY_asn1_get0(i);
+            if (ameth->pkey_flags & ASN1_PKEY_ALIAS)
+                continue;
+            pkey = d2i_PrivateKey(ameth->pkey_id, NULL, &blob, len);
```

Comment:
Not particularly, the ASN.1 definition of each is different enough for them to be recognisable.